### PR TITLE
Kcache flush

### PR DESCRIPTION
**Pull request by cosunae**

### CODE REVIEW

**Comment by cosunae** [comment]
launch pipeline

File: src/dawn/CodeGen/Cuda/CudaCodeGen.h [hunk 1]
```diff
@@ -35,6 +35,16 @@ class OptimizerContext;
 namespace codegen {
 namespace cuda {
 
+namespace impl_ {
```
Comment (twicki) [inline src/dawn/CodeGen/Cuda/CudaCodeGen.h]: what is the reasoning behind this namespacing here? Would it not be nicer if this was a private member of the class (as for example here:)
https://github.com/MeteoSwiss-APN/dawn/blob/74fa690bbe58016cc3a6bdf091e9b51ab89e78c7/src/dawn/IIR/StencilMetaInformation.h#L35

File: src/dawn/IIR/MultiStage.cpp [hunk 2]
```diff
@@ -129,6 +129,30 @@ iir::Cache& MultiStage::setCache(iir::Cache::CacheTypeKind type, iir::Cache::Cac
...
+  for(const auto& doMethod : iterateIIROver<DoMethod>(*this)) {
+    if(!doMethod->getFields().count(accessID)) {
+      continue;
+    }
+    const auto& interval = doMethod->getInterval();
+
+    if(loopOrder_ == LoopOrderKind::LK_Backward) {
+      if(interval.bound(Interval::Bound::lower) < level.bound() || !init) {
+        level = interval.lowerIntervalLevel();
+        init = true;
```
Comment (twicki) [inline src/dawn/IIR/MultiStage.cpp]: why not just return early if you only want to check once? And is this the intended behavior?

File: test/unit-test/dawn/Optimizer/Passes/samples/compute_extent_test_stencil_02_gen.cpp [hunk 3]
```diff
@@ -0,0 +1,224 @@
+// gtclang (0.0.1-a685a2b-x86_64-linux-gnu-7.3.0)
```
Comment (twicki) [inline test/unit-test/dawn/Optimizer/Passes/samples/compute_extent_test_stencil_02_gen.cpp]: this file should not be here, right?

File: src/dawn/CodeGen/Cuda/CodeGeneratorHelper.cpp [hunk 4]
```diff
@@ -32,6 +33,15 @@ std::string CodeGeneratorHelper::indexIteratorName(Array3i dims) {
   return n_;
 }
 
+bool CodeGeneratorHelper::useNormalIteratorForTmp(const std::unique_ptr<iir::MultiStage>& ms) {
+  for(const auto& stage : iterateIIROver<iir::Stage>(*ms)) {
```
[the remark this answers is not in the document]
Comment (cosunae) [reply src/dawn/CodeGen/Cuda/CodeGeneratorHelper.cpp]: done

File: src/dawn/CodeGen/Cuda/CacheProperties.cpp [hunk 5]
```diff
@@ -87,6 +87,10 @@ bool CacheProperties::requiresFill(const iir::Cache& cache) {
   return ((cache.getCacheIOPolicy() == iir::Cache::CacheIOPolicy::fill)); // ||
 }
 
+bool CacheProperties::requiresFlush(const iir::Cache& cache) {
+  return ((cache.getCacheIOPolicy() == iir::Cache::CacheIOPolicy::flush)); // ||
```
[the remark this answers is not in the document]
Comment (cosunae) [reply src/dawn/CodeGen/Cuda/CacheProperties.cpp]: done

File: src/dawn/CodeGen/Cuda/CodeGeneratorHelper.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (cosunae) [reply src/dawn/CodeGen/Cuda/CodeGeneratorHelper.cpp]: done

File: src/dawn/CodeGen/Cuda/CudaCodeGen.cpp [hunk 6]
```diff
@@ -569,6 +602,172 @@ void CudaCodeGen::generateFillKCaches(
...
+std::unordered_map<iir::Extents, std::vector<impl_::KCacheProperties>>
+CudaCodeGen::buildKCacheProperties(const std::unique_ptr<iir::MultiStage>& ms,
+                                   const iir::Interval& interval,
+                                   const CacheProperties& cacheProperties,
+                                   const bool checkStrictIntervalBound) const {
+
+  std::unordered_map<iir::Extents, std::vector<impl_::KCacheProperties>> kCacheProperty;
+  auto intervalFields = ms->computeFieldsAtInterval(interval);
+
+  for(const auto& cachePair : ms->getCaches()) {
```
[the remark this answers is not in the document]
Comment (cosunae) [reply src/dawn/CodeGen/Cuda/CudaCodeGen.cpp]: done

File: src/dawn/CodeGen/Cuda/CudaCodeGen.cpp [hunk 7]
```diff
@@ -680,8 +880,10 @@ void CudaCodeGen::generateKCacheSlide(MemberFunction& cudaKernel,
 }
 bool CudaCodeGen::intervalRequiresSync(const iir::Interval& interval, const iir::Stage& stage,
                                        const std::unique_ptr<iir::MultiStage>& ms) const {
-  // if the stage is the last stage, it will require a sync (to ensure we sync before the write of a
-  // previous stage at the next k level), but only if the stencil is not pure vertical and ij caches
+  // if the stage is the last stage, it will require a sync (to ensure we sync before the write of
+  // a
```
[the remark this answers is not in the document]
Comment (cosunae) [reply src/dawn/CodeGen/Cuda/CudaCodeGen.cpp]: fixed

File: src/dawn/IIR/MultiStage.cpp (hunk 2 above)
Comment (cosunae) [reply src/dawn/IIR/MultiStage.cpp]: this pattern is like computing a min, you might enter multiple times in the level =, which gets updated , so the early return would not work ehre

File: test/unit-test/dawn/Optimizer/Passes/samples/compute_extent_test_stencil_02_gen.cpp (hunk 3 above)
Comment (cosunae) [reply test/unit-test/dawn/Optimizer/Passes/samples/compute_extent_test_stencil_02_gen.cpp]: fixed

File: src/dawn/CodeGen/Cuda/CudaCodeGen.h (hunk 1 above)
Comment (cosunae) [reply src/dawn/CodeGen/Cuda/CudaCodeGen.h]: done

File: src/dawn/CodeGen/Cuda/CudaCodeGen.cpp [hunk 8]
```diff
@@ -107,7 +125,8 @@ void CudaCodeGen::generateKCacheDecl(MemberFunction& kernel,
 
     if(cache.getCacheType() != iir::Cache::CacheTypeKind::K ||
        ((cache.getCacheIOPolicy() != iir::Cache::CacheIOPolicy::local) &&
-        (cache.getCacheIOPolicy() != iir::Cache::CacheIOPolicy::fill)))
+        (cache.getCacheIOPolicy() != iir::Cache::CacheIOPolicy::fill) &&
```
[the remark this answers is not in the document]
Comment (cosunae) [reply src/dawn/CodeGen/Cuda/CudaCodeGen.cpp]: they all will be gone, once all the kcache policies are implemented

**Pull request closed without merging**